vllm/v1/worker/tpu_model_runner.py

```diff
@@ -0,0 +1,729 @@
+
+from vllm.attention import AttentionMetadata
+from vllm.config import VllmConfig
+from vllm.forward_context import set_forward_context
+from vllm.logger import init_logger
+from vllm.model_executor.model_loader import get_model
+from vllm.v1.attention.backends.pallas import (PallasAttentionBackend,
+                                               PallasMetadata)
+from vllm.v1.kv_cache_interface import FullAttentionSpec, KVCacheConfig
+from vllm.v1.outputs import ModelRunnerOutput
+from vllm.v1.utils import bind_kv_cache
+from vllm.v1.worker.gpu_input_batch import CachedRequestState, InputBatch
+from vllm.v1.worker.model_runner_base import ExecutionMode, ModelRunnerBase
+
+if TYPE_CHECKING:
+    from vllm.v1.core.scheduler import SchedulerOutput
+
+logger = init_logger(__name__)
+
+# Here we utilize the behavior that out-of-bound index is ignored.
+# FIXME(woosuk): Find a more reliable way to prevent possible bugs.
+_PAD_SLOT_ID = 1_000_000_000
+# FIXME(woosuk): Temporarily disabled top-p sampling since it's too slow.
+_ENABLE_TOP_P = False
+# FIXME(woosuk): A temporary hack to support `n > 1`.
+# This can significantly affect the performance if too large.
+_MAX_NUM_SAMPLES = 128
+
+
+@dataclass
+class PromptInputData:
+
+    req_ids: List
+    prompt_lens: List
+    input_tokens: List
+    input_positions: List
+    attn_metadata: List
+
+    def zipped(self):
+        return zip(self.req_ids, self.prompt_lens, self.input_tokens,
+                   self.input_positions, self.attn_metadata)
+
+
+@dataclass
+class DecodeInputData:
+    req_ids: List
+    input_tokens: Optional[torch.Tensor] = None
+    input_positions: Optional[torch.Tensor] = None
+    attn_metadata: Optional[PallasMetadata] = None
+
+
+class TPUModelRunner(ModelRunnerBase):
+
+    def __init__(
+        self,
+        vllm_config: VllmConfig,
+        device: torch.device,
+    ):
+        super().__init__(vllm_config, device)
+
+        # Persistent batch.
+        self.input_batch = InputBatch(
+            max_num_reqs=self.max_num_reqs,
+            max_model_len=self.max_model_len,
+            max_num_blocks_per_req=self.max_num_blocks_per_req,
+            device=self.device,
+            pin_memory=self.pin_memory,
+            vocab_size=self.model_config.get_vocab_size(),
+        )
+
+        # Request states.
+        self.requests: Dict[str, CachedRequestState] = {}
+
+        # KV caches for forward pass
+        self.kv_caches: List[Tuple[torch.Tensor, torch.Tensor]] = []
+
+        # Used to initialize positions for the individual prefills
+        self.prefill_input_positions = torch.tensor(range(self.max_model_len),
+                                                    device="cpu",
+                                                    dtype=torch.int32).reshape(
+                                                        1, -1)
+
+    def _prepare_prompt_inputs(
+        self,
+        scheduler_output: "SchedulerOutput",
+    ) -> PromptInputData:
```

Comment:
this is V0 style (`_prepare_prompt_inputs` + `_prepare_decode_inputs`), can we reuse 
```
def _prepare_inputs(self, scheduler_output: "SchedulerOutput"):
```
from V1 GPU model runner, as it prepare both prefill and decode inputs ?